sdk/src/main/scala/io/horizen/account/state/ForgerStakeStorageV3.scala

```diff
@@ -0,0 +1,404 @@
+               stakedAmount: BigInteger): Unit = {
+
+    val forgerKey = getForgerKey(signKey, vrfPublicKey)
+    val forgerHistory = StakeHistory(forgerKey)
+    val forgerHistorySize =  forgerHistory.getSize(view)
+    if (forgerHistorySize == 0)
+      throw new ExecutionRevertedException(s"Forger doesn't exist.")
+
+    forgerHistory.updateOrAddCheckpoint(view, forgerHistorySize, epochNumber, latestStake => latestStake.add(stakedAmount))
+
+    val stakeKey = getStakeKey(forgerKey, delegatorPublicKey)
+    val stakeHistory = StakeHistory(stakeKey)
+    val stakeHistorySize = stakeHistory.getSize(view)
+    stakeHistory.updateOrAddCheckpoint(view, stakeHistorySize, epochNumber, latestStake => latestStake.add(stakedAmount))
+    if (stakeHistorySize == 0)
+      addNewDelegator(view, forgerKey, delegatorPublicKey)
+  }
+
+  def removeStake(view: BaseAccountStateView,
+                  blockSignProposition: PublicKey25519Proposition,
+                  vrfPublicKey: VrfPublicKey,
+                  epochNumber: Int,
+                  delegatorPublicKey: Address,
+                  stakedAmount: BigInteger): Unit = {
+
+
+    val forgerKey = getForgerKey(blockSignProposition, vrfPublicKey)
+    val forgerHistory = StakeHistory(forgerKey)
+    val forgerHistorySize = forgerHistory.getSize(view)
+    if (forgerHistorySize == 0)
+      throw new ExecutionRevertedException("Forger doesn't exist.")
+
+    val stakeKey = getStakeKey(forgerKey, delegatorPublicKey)
+    val stakeHistory = StakeHistory(stakeKey)
+    val stakeHistorySize = stakeHistory.getSize(view)
+    if (stakeHistorySize == 0)
+      throw new ExecutionRevertedException(s"Delegator ${BytesUtils.toHexString(delegatorPublicKey.toBytes)} doesn't have stake with the forger.")
+
+
+    def subtractStake(latestStake: BigInteger): BigInteger = {
+      val newAmount = latestStake.subtract(stakedAmount)
+      if (newAmount.signum() == -1)
+        throw new ExecutionRevertedException(s"Not enough stake. Claimed $stakedAmount, available $latestStake")
+      newAmount
+    }
+
+    stakeHistory.updateOrAddCheckpoint(view, stakeHistorySize, epochNumber, subtractStake)
+    forgerHistory.updateOrAddCheckpoint(view, forgerHistorySize, epochNumber, subtractStake)
+  }
+
+  def getAllForgerStakes(view: BaseAccountStateView): Seq[(PublicKey25519Proposition, VrfPublicKey, Address, BigInteger)] = {
+    val listOfForgerKeys = ForgerMap.getKeys(view)
+    listOfForgerKeys.flatMap { forgerKey =>
+      val forger = ForgerMap(view, forgerKey)
+      val delegatorList = DelegatorList(forgerKey)
+      val delegatorSize = delegatorList.getSize(view)
+      val listOfStakes: ListBuffer[(PublicKey25519Proposition, VrfPublicKey, Address, BigInteger)] = ListBuffer()
+      for (idx <- 0 until delegatorSize) {
+        val delegator = delegatorList.getDelegatorAt(view, idx).address()
+        val stakeKey = getStakeKey(forgerKey, delegator)
+        val stakeHistory = StakeHistory(stakeKey)
```

Comment:
There's lots of places where we have these 2 lines one after another - why don't we combine it into a single method getStakeHistory that calculates key and returns history right away. Same for forgerHistory

Reply:
That's good point. I'll do it for stakeHistory. For forgerHistory, I still prefer like it is, because the forgerKey is used not only for getting the forgerHistory, but also for other operations in the same scope and I prefer not to calculate it more than once.